projects/flimnap/main.cc

```diff
@@ -176,29 +177,40 @@ namespace {
 
     // Transmission thread definitions
     constexpr std::size_t TX_MSG_BUFFER_SIZE = 16;
+    constexpr std::size_t PB_POSE_POOL_SIZE = 4;
     mailbox_t tx_mailbox;
-    std::array<msg_t, TX_MSG_BUFFER_SIZE> tx_msg_buffer;
+    msg_t tx_msg_buffer[TX_MSG_BUFFER_SIZE];
+    BicyclePoseMessage pb_pose_buffer[PB_POSE_POOL_SIZE] __attribute__((aligned(sizeof(stkalign_t))));
+    MEMORYPOOL_DECL(pb_pose_pool, sizeof(BicyclePoseMessage), nullptr);
 
     THD_WORKING_AREA(wa_transmission_thread, 2048);
     THD_FUNCTION(transmission_thread, arg) {
         (void)arg;
-        msg_t msg;
-        int32_t count;
-        static char serial_str[128];
 
         chRegSetThreadName("transmission");
-        chMBObjectInit(&tx_mailbox, tx_msg_buffer.data(), tx_msg_buffer.size());
+        chMBObjectInit(&tx_mailbox, tx_msg_buffer, TX_MSG_BUFFER_SIZE);
+        chPoolObjectInit(&pb_pose_pool, sizeof(BicyclePoseMessage), nullptr);
+        chPoolLoadArray(&pb_pose_pool, static_cast<void*>(pb_pose_buffer), PB_POSE_POOL_SIZE);
         while (true) {
-            chMBFetch(&tx_mailbox, &msg, TIME_INFINITE);
+            msg_t obj;
+            chMBFetch(&tx_mailbox, &obj, TIME_INFINITE);
             chSysLock();
-            count = chMBGetUsedCountI(&tx_mailbox);
+            int32_t count = chMBGetUsedCountI(&tx_mailbox);
             chSysUnlock();
 
             int formatted_bytes;
             static char serial_str[128];
             formatted_bytes = chsnprintf(serial_str, sizeof(serial_str)/sizeof(serial_str[0]),
-                    "[tx] current MB used count is %d, last msg: %d\r\n", count, msg);
-            usbTransmit(SDU1.config->usbp, SDU1.config->bulk_in, (const uint8_t*)serial_str, formatted_bytes);
+                    "[tx] current MB used count is %d, last msg: 0x%x\r\n", count, obj);
+
+            // TODO: encode pose to buffer? or should we have a received an encoded pose already?
+            // TODO: How to determine which pool the object belongs to? Compare address of pool?
+            if ((reinterpret_cast<BicyclePoseMessage*>(obj) >= pb_pose_buffer) &&
+                (reinterpret_cast<BicyclePoseMessage*>(obj) <= &pb_pose_buffer[PB_POSE_POOL_SIZE]) &&
+                (obj % sizeof(stkalign_t) == 0)) {
```

Comment:
`obj % sizeof(stkalign_t) == 0)` checks to make sure that the pointer is aligned correctly so the if statement essentially is to determine if `obj` is a pointer to a valid `BicyclePoseMessage`.

I need to distinguish messages coming from the pose thread vs the simulation thread.